envoy/server/factory_context.h

```diff
@@ -272,6 +286,32 @@ class FactoryContext : public virtual ListenerAccessLogFactoryContext {
    * @return Router::Context& a reference to the router context.
    */
   virtual Router::Context& routerContext() PURE;
+
+  /**
+   * Create an FilterConfigProviderPtr for a filter config. The config providers may share
+   * the underlying subscriptions to the filter config discovery service.
+   * @param config_source supplies the extension configuration source for the filter configs.
+   * @param filter_config_name the filter config resource name.
+   * @param last_filter_in_filter_chain indicates whether this filter is the last filter in the
+   * configured chain
+   * @param filter_chain_type is the filter chain type
+   * @param listener_filter_matcher is the filter matcher for TCP listener filter. nullptr for other
+   * filter types.
+   *
+   * @return HttpExtensionConfigProvider
+   */
+  virtual Configuration::HttpExtensionConfigProvider createDynamicFilterConfigProvider(
+      const envoy::config::core::v3::ExtensionConfigSource& config_source,
+      const std::string& filter_config_name, bool last_filter_in_filter_chain,
+      const std::string& filter_chain_type,
+      const Network::ListenerFilterMatcherSharedPtr& listener_filter_matcher) PURE;
+
+  /**
+   * Returns the downstream filter config provider manager.
```

Comment:
Should this be in the server factory context maybe? Is that possible? ECDS is meant to be shared between listeners, so having a singleton attached to a listener might lead to bugs where we have two instances of the config providers. I've tried to find where the singleton instance is now - and it's not trivial to find in this PR.

Reply:
It is part of FactoryContextImpl. I believe it is across listeners